pkg/updater/download.go

```diff
@@ -116,41 +193,142 @@ func extractTarGz(archivePath string, destinationPath string) error {
 		default:
 			log.Warnf("Unsupported tar entry type %d for %s", header.Typeflag, header.Name)
 		}
 	}
+
+	// Process tar links afterwards as they may depend on other files being written
+	err = processTarLinks(0, destinationPath, tarLinks)
+	if err != nil {
+		return err // already wrapped
+	}
+
 	log.Debugf("Successfully extracted archive %s to %s", archivePath, destinationPath)
 	return nil
 }
 
-// extractTarGzFile extracts a file from a tar.gz archive.
+// extractTarFile extracts a file from a tar archive.
 // It is separated from extractTarGz to ensure `defer f.Close()` is called right after the file is written.
-func extractTarGzFile(targetPath string, reader io.Reader) error {
-	err := os.MkdirAll(filepath.Dir(targetPath), 0755)
+func extractTarFile(targetPath string, reader io.Reader) error {
+	err := os.MkdirAll(filepath.Dir(targetPath), 0o755)
 	if err != nil {
 		return fmt.Errorf("could not create directory: %w", err)
 	}
-	f, err := os.Create(targetPath)
+	f, err := os.Create(filepath.FromSlash(targetPath))
 	if err != nil {
 		return fmt.Errorf("could not create file: %w", err)
 	}
 	defer f.Close()
-	_, err = io.Copy(f, reader)
+
+	n, err := io.Copy(f, io.LimitReader(reader, maxArchiveFileSize))
 	if err != nil {
+		if errors.Is(err, io.EOF) && n == maxArchiveFileSize {
+			defer func() {
+				if err := os.Remove(targetPath); err != nil {
+					log.Errorf("Could not remove truncated file %q: %v", targetPath, err)
+				} else {
+					log.Debug("Removing truncated file %q", targetPath)
+				}
+			}()
+			return fmt.Errorf("content truncated: file %q is too large: %w", targetPath, err)
+		}
 		return fmt.Errorf("could not write file: %w", err)
 	}
 	return nil
 }
+
+// processTarLinks processes the symlinks in a tar archive and makes sure the destinations exist.
+// It is called recursively to handle symlinks that depend on other symlinks as we
+// iteratively create them.
+func processTarLinks(level int, destinationPath string, symlinks []*tar.Header) error {
+	if len(symlinks) == 0 {
+		// Fast path
+		return nil
+	}
+
+	// Check recursion level
+	if level >= maxArchiveLinkDepth {
+		return fmt.Errorf("maximum symlink recursion level reached (%d)", level)
+	}
+
+	// Prepare next pass
+	next := make([]*tar.Header, 0)
+
+	for _, hdr := range symlinks {
+		if hdr == nil {
+			continue
+		}
+
+		targetName := filepath.Join(destinationPath, hdr.Name)
+
+		// Generate the target link name depending on whether it is absolute
+		// or relative to the current file. Hardlinks are always absolute
+		var targetLinkName string
+		if filepath.IsAbs(hdr.Linkname) || hdr.Typeflag == tar.TypeLink {
+			targetLinkName = filepath.Join(destinationPath, hdr.Linkname)
+		} else {
+			targetLinkName = filepath.Join(filepath.Dir(targetName), hdr.Linkname)
+		}
+
+		// Check for zip-slip attacks
+		if !strings.HasPrefix(filepath.Join(destinationPath, targetLinkName), filepath.Clean(destinationPath)+string(os.PathSeparator)) {
```

Comment:
`targetLinkName` already contains the `destinationPath` , no need to re-append it.